discovery_engine_core/core/src/stack/filters/semantic.rs

```diff
@@ -220,9 +222,17 @@ pub(crate) fn filter_semantically(
         Criterion::MaxClusters(max_clusters) => find_n_clusters(&dendrogram, max_clusters),
     };
 
+    // among documents with the same label, keep the one with heaviest source weight
     izip!(documents, labels)
-        .unique_by(|(_, label)| *label)
-        .map(|(document, _)| document)
+        .fold(HashMap::new(), |mut map, (doc, label)| {
+            map.entry(label).or_insert_with(Vec::new).push(doc);
```

Comment:
:art: we could use the default right away
```suggestion
            map.entry(label).or_default().push(doc);
```